File: solution/cpp/144_BinaryTreePreorderTraversal.cpp
#include <iostream>
#include <vector>
#include <myTreeNode.h>
#include <myVector.h>
#include <myutils.h>

using namespace std;

class Solution {
public:
    vector<int> preorderTraversal(TreeNode *root) {
        vector<int> res;
        preorderTraversalUtils(root, res);
        return res;
    }

private:
    void preorderTraversalUtils(TreeNode *root, vector<int> &res) {
        if (root == nullptr)
            return;
        res.push_back(root->val);
        preorderTraversalUtils(root->left, res);
        preorderTraversalUtils(root->right, res);
    }
};

int main() {
//       1
//     /  \
//    2   3
//  /  \
// 4   5
    auto *so = new Solution();
    TreeNode *root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);
    vector<int> res = so->preorderTraversal(root);
    print_1d_vector(res);
    delete so;
    return 0;
}

File: solution/python/144_BinaryTreePreorderTraversal.py
from utils import *


class Solution:
    def preorderTraversal(self, root):
        res = []
        stack = []
        curr = root
        while curr is not None or len(stack) > 0:
            while curr is not None:
                stack.append(curr)
                res.append(curr.val)
                curr = curr.left
            curr = stack.pop()
            curr = curr.right
        return res


if __name__ == "__main__":
    root = TreeNode(val=1)
    root.right = TreeNode(val=2)
    root.right.left = TreeNode(val=3)
    print(root)
    s = Solution()
    print(s.preorderTraversal(root))
